### ISSUE TITLE
Run git fetches and post installs in parallel

### ISSUE THREAD

**Issue by repomaa** [crystal-lang/shards#239]
if possible run git fetches and post installs in parallel especially for running tests in a ci this would be helpful as the dependency fetching and building can take some time

**Comment by ysbaddaden** [comment]
I'm not sure we could make this possible. Quite the opposite, sadly :disappointed: 

Building can require transitive dependencies to be installed (see #124). In fact it could even require a such a dependency to have been built (:fearful:). Running the postinstall scripts shall be postponed to *after* we installed *all* dependencies. We could build a tree of dependencies, and try to install them in some order, but still, that could only happen *after* we successfully resolved the dependency tree, and thus *after* we cloned/fetched all repositories.

A solution to speed up CI/CD is to cache the `lib` folder, using a checksum of `shard.lock` in the cached filename to decide whether it should install+save the cache or restore it.

**Comment by bcardiff** [comment]
@jreinert the article at https://crystal-lang.org/2018/09/04/using-circleci-2.0-for-your-crystal-projects.html covers the how to cache the downloaded shards in CircleCI. Maybe that helps in your use case.

Git fetches could be run in parallel probably, like when `bundle --jobs` [was introduces](https://bundler.io/v1.5/whats_new.html)

**Comment by oz** [comment]
Does this mean that we could fetch repositories in parallel, and then process post-install sequentially? :thinking: 

CI caches are a good work-around for CI environments, but it would be a nice-to-have for others environments.

**Comment by RX14** [comment]
@ysbaddaden surely the best way would be to always cache lib and run `shards` again every time, then optimize shards to handle the up-to-date case really quickly.

**Comment by RX14** [comment]
I think that some of the cloning could be done in parallel though, because even the root shards.yml gives multiple dependencies which gives some immediate parallelism. After that attempt to create a DAG of dependencies and run the scripts bottom-up in parallel.

**Comment by ysbaddaden** [comment]
I think I misunderstood. This is about running clones/fetches in parallel, then running postinstall scripts in parallel. This isn't about running fetch/postinstall in parallel, right?

This could be possible, but would require some non negligible architectural changes in the implementation to be more actor-like and modular, than it is today (purely sequential).

**Comment by straight-shoota** [comment]
Closing as duplicate of #539

The aspect of running postinstall concurrently isn't covered there, but that doesn't seem very feasible to me. Anyway, a new issue should be opened for that.
